### ISSUE TITLE
Alternative to Windows Store for Windows

### ISSUE THREAD

**Issue by NicholasWMRitchie** [JuliaLang/juliaup#343]
My work computer is forbidden by policy to access the Windows Store.  Could there be an alternative source for JuliaUp for Windows as there is for Linux and Mac?

**Comment by davidanthoff** [comment]
@NicholasWMRitchie Could you try a new install option that avoids the store? Here is how:
- Click on https://install.julialang.org/Julia.appinstaller
- Launch the file that gets downloaded (just "open" it or double click on it in explorer)

Be warned, the version that you are getting there is very much a dev version, and you probably will have to uninstall this and then re-install a more robust version down the road. Right now I'm just trying to figure out whether this option is one that could solve this scenario for us.

**Comment by NicholasWMRitchie** [comment]
I did as you instructed.   It installs a 64-bit instance of Julia 1.7.3 and the instance works as expected.
I could not discover any way to install different Julia versions as the more complete Windows Store option does.

**Comment by lungben** [comment]
Thanks, this is really helpful!

I suggest to add the link to the appinstaller to the Readme file of this repo so that it is easier to find for other users.

**Comment by davidanthoff** [comment]
> I suggest to add the link to the appinstaller to the Readme file of this repo so that it is easier to find for other users.

For now it is an experiment, as soon as we decide that we want to have that option permantely I'll put a link in the README. Oh, and we need to change code signing from my personal cert to the JuliaComputing one as well before we really declare this _done_.

**Comment by NicholasWMRitchie** [comment]
Is there a reason that the "juliaup self update" is not implemented on this version or is this just one of those things that you still need to complete?

I just reinstalled Julia.appinstaller to get the latest 1.8.0-rc4 version as it wasn't available on the initial test version.

**Comment by davidanthoff** [comment]
> Is there a reason that the "juliaup self update" is not implemented on this version or is this just one of those things that you still need to complete?

Yep, that still needs to be implemented. I think in theory this appinstaller distributed version should auto-check for new versions, and then auto-update in the background, but there might be delays in that. Once I implement the `juliaup self update` thing that should force-trigger that.

I also plan to change how Juliaup in general learns about new Julia versions, at that point the delays should be minimal.

**Comment by bilderbuchi** [comment]
>  I think in theory this appinstaller distributed version should auto-check for new versions, and then auto-update in the background, but there might be delays in that. 

Is there some way to detect/confirm that? I installed the appinstaller 1.7.20 version a couple of days ago -- worked flawlessly on a machine without admin rights or a useful Windows Store. However, even after a reboot juliaup is still at that version (and therefore does not see the julia 1.8.0 release).

**Comment by oheil** [comment]
Just did an update on my work PC via Julia.appinstaller 
For me this is the currently the most convenient way of getting/updating Julia on Windows. Typically I don't have any problems with other ways of installing Julia, but for the general Julia users my guess is, this is the best, even in this early stage.
As I don't have any issues which aren't already known or anything unexpected this is just meant as an encouraging note and a general WELL DONE! Thank you for your great work.

**Comment by DilumAluthge** [comment]
@ViralBShah Can you give this approach a try?

**Comment by MatthijsBlom** [comment]
@davidanthoff I tried the Julia.appinstaller as well, and this is the result (immediate uppon clicking <kbd>Install</kbd>):

<details><summary>😢 Screenshot</summary>

![image](https://user-images.githubusercontent.com/19817960/195890218-295b0175-6a56-463c-8fb9-8f8df0aefe70.png)
</details>

**Comment by davidanthoff** [comment]
What version of Windows are you on? I thought MS had resolved this now (in a previous version of App installer this always happened...).

**Comment by oheil** [comment]
Not sure if this leads to a solution but @MatthijsBlom  can you open a PowerShell and show the output of
`Get-ExecutionPolicy -List`

**Comment by MatthijsBlom** [comment]
```
        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
  CurrentUser       Undefined
 LocalMachine    RemoteSigned
```

**Comment by oheil** [comment]
OK, it's same for me but I can install. So it was a wrong track, never mind.

**Comment by el-oso** [comment]
I tried on my company computer and I got the following message:
`App installation failed with error message: Appinstaller operation failed with error code 0x80070490. Detail: Element not found. (0x80070490)`

**Comment by gruemelmonster** [comment]
First I got the same error as MatthijsBlom but then I updated the Store App from inside the Store and now it worked fine.
(We can only install a bunch of apps from the store and sadly Julia is not (yet) one of them)

**Comment by MatthijsBlom** [comment]
I can confirm that updating "App Installer" fixes my previous problem.

**Comment by el-oso** [comment]
Not sure if this app was update since, but it is still not working (see screenshot).
![Capture](https://user-images.githubusercontent.com/15278831/209649466-217eddc5-7873-43bf-b139-9c48361d84e4.PNG)

**Comment by el-oso** [comment]
Version 1.8.16 was successfully installed on my work computer.

**Comment by phlavenk** [comment]
Does anybody have a recipe for how to install the juliaup on the Windows Server (2019 or 2022)? We are using the server editions on the SuperMicro stations (2 NUMA nodes, 100+ cores, 1.5TB RAM) but the .appinstaller file cannot be opened there.
I'd be glad to have juliaup (even only locally installed) on the system for the ease of use of the version multiplexing, even sacrificing the auto-update functionality.

**Comment by el-oso** [comment]
not really a show stopper, but a mild inconvenience. It seems that now I lost the ability to run julia in Windows from bash. 
Now I get an error message like:
`bash: /c/Users/USERNAME/AppData/Local/Microsoft/WindowsApp/julia: Permission denied` 

Not really a show stopper, but extended the usability of `shell>` mode from Windows. Otherwise, shell mode is sort of useless in Windows. If someone knows how to fix this problem, I would be very grateful :-)

**Comment by DilumAluthge** [comment]
> not really a show stopper, but a mild inconvenience. It seems that now I lost the ability to run julia in Windows from bash.

This seems like a pretty big bug to me. I frequently use Bash on Windows, and I would definitely want to be able to run Julia in Bash.

@davidanthoff @staticfloat Any ideas what might be going on?

**Comment by staticfloat** [comment]
Sounds like the file doesn't have executable permissions.  Try `chmod +x ./julia` from `bash`?

**Comment by LilithHafner** [comment]
> Oh, and we need to change code signing from my personal cert to the JuliaComputing one as well before we really declare this _done_.

OT: Is there a JuliaLang cert we could use instead?

**Comment by DilumAluthge** [comment]
> Sounds like the file doesn't have executable permissions. Try `chmod +x ./julia` from `bash`?

Even if this works, shouldn't Juliaup automatically set the correct permissions?

**Comment by el-oso** [comment]
Found a workaround that is not too bad type `winpty julia` or `winpty juliaup` and that works fine. Those can be made as an alias in your `.bashrc`

**Comment by ebenitez78** [comment]
I do not have access to Windows Store (employer police) thus I wanted to try this way.  But I do not have AppInstaller either, and as far as I know, I can install it only from the Windows Store.  Is there a third way to install juliaup?

**Comment by davidanthoff** [comment]
https://www.augmentedmind.de/2021/05/30/distributing-windows-applications/ is useful.

**Comment by dvetsch75** [comment]
> I do not have access to Windows Store (employer police) thus I wanted to try this way. But I do not have AppInstaller either, and as far as I know, I can install it only from the Windows Store. Is there a third way to install juliaup?

I am in the same situation.

**Comment by davidanthoff** [comment]
Just FYI, I'm working on a traditional MSI installer as another alternative.

**Comment by davidanthoff** [comment]
I have a new MSI installer option for situations where the Windows Store and App Installer don't work on Windows. Would be great if some folks could try this! Be warned, this is experimental/prerelease at the moment, though.

64 bit version: https://julialang-s3.julialang.org/juliaup/winmsi/Julia-v1.11.18-x64.msi
32 bit version: https://julialang-s3.julialang.org/juliaup/winmsi/Julia-v1.11.18-x86.msi

By default this will install per-user, so not require admin permissions. You can also install systemwide by running:
```
msiexec /i Julia-v1.11.18-x64.msi ALLUSERS=1
```
The main drawback of this is that there is no auto-update for Juliaup itself. To update Juliaup, one will have to download a new version of the MSI installer and then run that. I might be able to later automate this a little bit more, but I don't think we'll ever be able to get the complete silent auto-update feature we have everywhere else. But hey, this is meant as a last-resort option for folks on locked down servers etc, so hopefully this is ok.

Oh, and what this is installer does is: 1) drop the binaries onto the user system, 2) add a start menu shortcut and 3) add things to the `PATH`.

In any case, test reports would be great!

**Comment by dvetsch75** [comment]
> I have a new MSI installer option for situations where the Windows Store and App Installer don't work on Windows. Would be great if some folks could try this! Be warned, this is experimental/prerelease at the moment, though.


This has been working for me with no issues for a few weeks.

**Comment by gitboy16** [comment]
Hi, I was just wondering is there a portable version of juliaup?

**Comment by LilithHafner** [comment]
A cross-platform fat binary that runs `curl -fsSL https://install.julialang.org | sh` (or a Windows equivalent) seems like a reasonable file to distribute.

**Comment by davidanthoff** [comment]
> I was just wondering is there a portable version of juliaup?

Turns out I'm compiling one already as part of the build, just not putting it anywhere :) I actually have a need for this myself, so I'll probably start to attach those to the GitHub releases. That won't have any auto-update mechanism, though (of itself, it will of course update Julia versions).

**Comment by davidanthoff** [comment]
https://github.com/JuliaLang/juliaup/releases/tag/v1.12.5 now has portable binaries. I haven't tried them, though :)

**Comment by GlenHertz** [comment]
The App Installer doesn't work for me on a somewhat locked down Windows 11 machine.  I get the error:

`App installation failed with error message: The package full name returned from the AppxManifest (JuliaComputingInc.Julia_1.17.1.0_neutral_~_5z4q23t4ga8jg) does not match the name generated from the AppInstaller`

**Comment by davidanthoff** [comment]
@GlenHertz  Ah, that might actually be related to a change in certificate we recently had, and I probably have to update more things. I'll check right now.

**Comment by gbenezer** [comment]
Currently running the MSIX App Installer on my system, and it worked thank goodness! Been trying for a while to install juliaup.
Thank you for developing an alternative!

julia> versioninfo()
Julia Version 1.11.2
Commit 5e9a32e7af (2024-12-01 20:02 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 16 × AMD Ryzen 9 5900HX with Radeon Graphics
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, znver3)
Threads: 1 default, 0 interactive, 1 GC (on 16 virtual cores)

**Comment by rgaudette** [comment]
Using the Julia.Appinstaller version 1.17.11.0, I get the follow error:

App installation failed with error message: error 0x80070005: Windows cannot create the AppContainer profile for the JuliaComputingInc.Julia_1.17.11.0_x64__5z4q23t4ga8jg package. (0x80070005)

When I tried to install juliaup using `winget install julia -s msstore` I get the following error

Starting package install...
  ███████████████████████████     90%
Failed to install or upgrade Microsoft Store package. Error code: 0x80070005

The windows store rarely works on this machine and I have never been able to figure out what is wrong with it, so take this report with a grain of salt. Let me know if I can get you any more information.

**Comment by BambOoxX** [comment]
Hi, I just tried the `Julia.Appinsteller` and I get an installation failure, because the program could not install dependencies.
The computer is behind a proxy. Could it be that the installer fails to access dependencies because of a proxy environment variable ?

**Comment by BambOoxX** [comment]
@davidanthoff I just tried again the installer in v1.19.3.0 but that still fails do you have any ideas ?

**Comment by davidanthoff** [comment]
Yes, app installer is also trying to install some dependencies. Maybe try the MSI option next?

**Comment by BambOoxX** [comment]
Ah yes, this seems to work indeed ! When migrating to this install of julia(up) should I remove my `.julia` folder to reset everything ?

**Comment by davidanthoff** [comment]
No, you can leave the `.julia` folder as is.

**Comment by demianelnecave** [comment]
Hi! I installed it succesfully, but when I put      `julia` or `juliaup` in the command-line it doesn't recognise the commands. I can open the Julia interface by typing it in the Windows search bar.

**Comment by BambOoxX** [comment]
FYI, I just tested the Appinstaller v1.22 and it works properly even with MSStore locked. It's the first version of the appinstaller I tried that worked. Thanks for the hard work

**Comment by davidanthoff** [comment]
Awesome!

**Comment by lwhitefox** [comment]
The installer worked just fine on company machine where the store access is blocked. Thanks for setting up
